### PR TITLE
macros 1.1: future proofing and cleanup

### PR DESCRIPTION
**Pull request by jseyfried**
This PR
- uses the macro namespace for custom derives (instead of a dedicated custom derive namespace),
- relaxes the shadowing rules for `#[macro_use]`-imported custom derives to match the shadowing rules for ordinary `#[macro_use]`-imported macros, and
- treats custom derive `extern crate`s like empty modules so that we can eventually allow, for example, `extern crate serde_derive; use serde_derive::Serialize;` backwards compatibly.

r? @alexcrichton

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @nrc

File: src/test/compile-fail-fulldeps/proc-macro/shadow.rs [hunk 1]
```diff
@@ -9,13 +9,12 @@
...
 // aux-build:derive-a.rs
-// aux-build:derive-a-2.rs
 
 #![feature(proc_macro)]
 
 #[macro_use]
 extern crate derive_a;
 #[macro_use]
-extern crate derive_a_2; //~ ERROR: cannot shadow existing derive mode `A`
+extern crate derive_a; //~ ERROR `derive_a` has already been defined
```
Comment (alexcrichton) [inline src/test/compile-fail-fulldeps/proc-macro/shadow.rs]: Hm isn't this testing something else than before? Before it was preventing two `#[proc_macro_derive(A)]` was an error, but here it's erroring out on the `extern crate` annotations?

**Review by alexcrichton — COMMENTED** [review]
Looks good to me but I'm not 100% familiar with all the changes, so 

r? @nrc

File: src/test/compile-fail-fulldeps/proc-macro/shadow.rs (hunk 1 above)
Comment (jseyfried) [inline src/test/compile-fail-fulldeps/proc-macro/shadow.rs]: Yeah, it used to test that a `#[macro_use]`-imported custom derive cannot shadow an existing `#[macro_use]`-imported custom derive.

This PR changes `#[macro_use]`-imported custom derives to have the same scope and shadowing restrictions as other `#[macro_use]`-imported macros, so it is no longer an error for an unexpanded `#[macro_use]` to import a shadowing custom derive (just as it is not an error ordinary unexpanded `#[macro_use]`s to shadow).

**Pull request merged**